src/Juvix/Data/Effect/Files/Pure.hs

```diff
@@ -0,0 +1,141 @@
+module Juvix.Data.Effect.Files.Pure where
+
+import Data.HashMap.Strict qualified as HashMap
+import Data.Tree
+import Juvix.Data.Effect.Files.Base
+import Juvix.Prelude.Base
+import Path hiding ((</>))
+import Path qualified
+import Prelude qualified
+
+data FS = FS
+  { _fsRoot :: Path Rel Dir,
+    _fsNode :: FSNode
+  }
+
+data FSNode = FSNode
+  { _dirFiles :: HashMap (Path Rel File) Text,
+    _dirDirs :: HashMap (Path Rel Dir) FSNode
+  }
+
+emptyNode :: FSNode
+emptyNode =
+  FSNode
+    { _dirFiles = mempty,
+      _dirDirs = mempty
+    }
+
+makeLenses ''FS
+makeLenses ''FSNode
+
+-- | Files should have a common root
+mkFS :: HashMap FilePath Text -> FS
+mkFS tbl = case (HashMap.toList (rootNode ^. dirDirs), toList (rootNode ^. dirFiles)) of
+  ([(r, d')], []) -> FS r d'
+  _ -> impossible
+  where
+    rootNode :: FSNode
+    rootNode = HashMap.foldlWithKey' insertFile emptyNode tbl
+    insertFile :: FSNode -> FilePath -> Text -> FSNode
+    insertFile dir0 fp contents = go (splitPath fp) dir0
+      where
+        go :: [FilePath] -> FSNode -> FSNode
+        go l dir = case l of
+          [] -> impossible
+          [f] -> set (dirFiles . at (relFile f)) (Just contents) dir
+          (d : ds) -> over dirDirs (HashMap.alter (Just . helper) (relDir d)) dir
+            where
+              helper :: Maybe FSNode -> FSNode
+              helper = maybe (helper (Just emptyNode)) (go ds)
+
+relFile :: FilePath -> Path Rel File
+relFile = fromJust . parseRelFile
+
+relDir :: FilePath -> Path Rel Dir
+relDir = fromJust . parseRelDir
+
+destructPath :: Path b Dir -> [Path Rel Dir]
```

Comment:
suggestion: Add `destructPath` and `destructFilePath` to Juvix.Prelude

Reply:
I thought about that but I wasn't sure since we are still using Filepath in most of the codebase.

I think I'll create a `Juvix.Prelude.Path` file and move these functions there. For now `Juvix.Prelude` will not export that

Reply:
I agree with this reasoning. Putting these functions in `Juvix.Prelude.Path` means we're more likely to remember that they exist if we ever move over to `path`.